src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp

```diff
@@ -289,138 +289,97 @@ void ShenandoahBarrierSetAssembler::load_reference_barrier_not_null(MacroAssembl
 
   Address gc_state(thread, in_bytes(ShenandoahThreadLocalData::gc_state_offset()));
   __ testb(gc_state, ShenandoahHeap::HAS_FORWARDED);
-  __ jccb(Assembler::zero, done);
-
-  // Use rsi for src address
-  const Register src_addr = rsi;
-  // Setup address parameter first, if it does not clobber oop in dst
-  bool need_addr_setup = (src_addr != dst);
-
-  if (need_addr_setup) {
-    __ push(src_addr);
-    __ lea(src_addr, src);
-
-    if (dst != rax) {
-      // Move obj into rax and save rax
-      __ push(rax);
-      __ movptr(rax, dst);
-    }
-  } else {
-    // dst == rsi
-    __ push(rax);
-    __ movptr(rax, dst);
-
-    // we can clobber it, since it is outgoing register
-    __ lea(src_addr, src);
-  }
-
-  save_xmm_registers(masm);
-  __ call(RuntimeAddress(CAST_FROM_FN_PTR(address, ShenandoahBarrierSetAssembler::shenandoah_lrb())));
-  restore_xmm_registers(masm);
-
-  if (need_addr_setup) {
-    if (dst != rax) {
-      __ movptr(dst, rax);
-      __ pop(rax);
+  __ jcc(Assembler::zero, heap_stable);
+
+  Register tmp1 = noreg;
+  if (!weak) {
+    // Test for object in cset
+    // Allocate tmp-reg.
+    for (int i = 0; i < 8 && !tmp1->is_valid(); i++) {
+      Register r = as_Register(i);
+      if (r != rsp && r != rbp && r != dst && r != src.base() && r != src.index()) {
+        tmp1 = r;
```

Comment:
Minor nit: looks like you just want to `break` from here, and not check `tmp1->is_valid()` in the loop predicate?